torchaudio/functional/functional.py

```diff
@@ -1353,6 +1369,34 @@ def _get_sinc_resample_kernel(
     return torch.stack(kernels).view(new_freq, 1, -1).mul_(scale), width
 
 
+def _apply_sinc_resample_kernel(
+        waveform: Tensor,
+        orig_freq: float,
+        new_freq: float,
+        gcd: int,
+        kernel: Tensor,
+        width: int,
+):
+    orig_freq = int(orig_freq) // gcd
+    new_freq = int(new_freq) // gcd
+
+    # pack batch
+    shape = waveform.size()
+    waveform = waveform.view(-1, shape[-1])
+    kernel = kernel.to(device=waveform.device, dtype=waveform.dtype)
```

Comment:
Considering the fact that this function will be called from Transform, we might want to avoid moving parameter to a certain device.

In a use case like 
1. Initialize `Resample`
2. Move the pipeline to a specific device
3. Run a inference.

and let's say that the pipeline is in GPU but the waveform is on CPU. We want the pipeline to fail so that users can fix the pipeline, instead of performing the operation on CPU, which incurs the cost of moving kernel from GPU to CPU. (well and then it will fail in the next step of pipeline, which expects the Tensor to be on GPU)

I think, moving the `kernel` to target device can happen in between `_get_kernel` function and `_apply_kernel` functions.

Reply:
sure, where do you think target device/dtype should initially be set in transforms?

Reply:
> sure, where do you think target device/dtype should initially be set in transforms?

They are expected to be set by users with `.to`, so we do not need to set it explicitly.

```python
resamp = T.Resample(...)
resamp.to(device, dtype)
```

Reply:
as discussed offline, this is BC-breaking and left as a follow-up